### PR TITLE
[Core] NURBS curve on surface

### PR DESCRIPTION
**Pull request by andreas-apostolatos**
Hi all,

I am adding here the nurbs curve on surface class which is used for defining curves in the parametric space of surfaces, such as trimming curves.

Thanks!

### CODE REVIEW

File: kratos/geometries/nurbs_curve_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_geometry.h]: You could do this in one line

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]: void GlobalSpaceCoordinates(...) const override

and maybe put it after the GlobalCoordinates to have the same order as in the master. In #5703 are made the same restructuring as we should do here

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]: as we found another way in the projection and tesselation you can derive directly from geometry

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]: you can copy from #5703

**Review by tteschemacher — COMMENTED** [review]
some comments adapting the updated PRs

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]: Change please to the new interface

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]: override

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (pooyan-dadvand) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]: Be careful that now your input and output are the same. So the use of noalias in the GlobalCoordiantes method would give wrong results.

**Review by pooyan-dadvand — COMMENTED** [review]
I would double-check the aliasing. And is good to use noalias in operations and having the in/out variables separate to improve performance

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (andreas-apostolatos) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h]: +1 @pooyan-dadvand thanks for the observation!!

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h:159]: I'm not sure if it's good to return to this value and pass the rResult.

I would define the variable before and then only pass it by reference.

Comment (philbucher) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:159]: I agree, think this is legacy, not sure why it was done in the first place

Mybe @pooyan-dadvand has a comment

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (philbucher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h:254]: does this have to be a shared ptr?
(@tteschemacher )

Comment (tteschemacher) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:254]: Not necessarily, but in this case it does not really make a difference.

If there will be constructors that construct e.g. the curve then it needs to be a smart pointer.

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (andreas-apostolatos) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:159]: I will write it on a local variable then

**Comment by andreas-apostolatos** [comment]
Hello together!

can we proceed with approving the PR if there are no more comments? Sorry for pressuring :)

Cheers!

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:159]: After that change I we can merge it.

Comment (pooyan-dadvand) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:159]: Yes, it is a legacy design and my fault! :see_no_evil: 

Meanwhile, I don't understand why you need `result_local`?

Comment (andreas-apostolatos) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:159]: The idea is not to mix rResult with coming from mpNurbsCurve->GlobalCoordinates with the rResult coming as an input, cause they both need to be fed into mpNurbsSurface->GlobalCoordinates

**Comment by andreas-apostolatos** [comment]
Hello again,

twenty days have passed since my last request regarding a review, I am sorry for being pressuring but can you please proceed with approving this pull request or give me some feedback as to whether something needs to be changed?

Looking forward to some action in this pull request, it is stagnating and needs attention! Thanks in advance

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h:135]: This create function is required.

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]: weird name

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (andreas-apostolatos) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:135]: This function expects as arguments PointsArrayType const& ThisPoints to create the geometry but for this type of geometry I don't know which points to pass, the ones of the surface or the ones of the curve? And if it is those of the curve, the ones in the parametric or the ones the physical space? Need opinions!

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]: I would go here for the nodes of the Surface. Because, essentially those are the one which are used as the dofs. Might be problematical if it is the other way round.

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (tteschemacher) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:135]: The PointsArrayType is defined by the BaseType. This is essentially the surface points. For you, here it does not make a difference.

File: kratos/geometries/nurbs_curve_on_surface_geometry.h
Comment (andreas-apostolatos) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h]: Thanks for the input @tteschemacher. I am trying to get it fixed now, but I don't think I can create the necessary shared pointer only passing as argument,

TSurfaceContainerPointType const& ThisPoints

cause this class expects in its constructor a pointer to a surface and a pointer to a curve. Am I getting it right? The following code,

typename BaseType::Pointer Create(
        TSurfaceContainerPointType const& ThisPoints) const override
    {
        return Kratos::make_shared<NurbsCurveOnSurfaceGeometry>(ThisPoints);
    }

gives me the following compilation error,

error: no matching function for call to ‘Kratos::NurbsCurveOnSurfaceGeometry<3, Kratos::PointerVector<Kratos::Point>, Kratos::PointerVector<Kratos::Point> >::NurbsCurveOnSurfaceGeometry(const Kratos::PointerVector<Kratos::Point>&)’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }

Looking forward to some enlightening. Thanks!

**Review by tteschemacher — APPROVED** [review]
Thanks @andreas-apostolatos 

The create function is not urgent and will be addressed in a new PR.

Further the class was discussed deeply and follow all criteria. Thus, I am approving it that we'll be able to proceed here.

**Pull request merged**
